sortthread.go

```diff
@@ -57,3 +62,112 @@ type Thread struct {
 	Id       uint32
 	Children []*Thread
 }
+
+var (
+	tabsContinuation = regexp.MustCompile(`[\t\\]`)
+	repeatedSpaces   = regexp.MustCompile("[ ]+")
+
+	// Includes regex for ABNF rules relevant to base subject
+	// Note that all ABNF strings are considered lowercase
+	subjFwdHeader = "[fwd:"
+	subjFwdTrail  = "]"
+
+	// BLOBCHAR        = %x01-5a / %x5c / %x5e-ff
+	// subj-blob       = "[" *BLOBCHAR "]" *WSP
+	subjBlob       = `\[\x01-\x5a\x5c\x5e-\xff]\]\s*`
+	subjBlobPrefix = regexp.MustCompile(fmt.Sprintf("^%s", subjBlob))
+
+	// subj-refwd      = ("re" / ("fw" ["d"])) *WSP [subj-blob] ":"
+	subjReFwd = fmt.Sprintf(`(?:(?:re)|(?:fwd?))\s*(?:%s)?:`, subjBlob)
+	// subj-leader     = (*subj-blob subj-refwd) / WSP
+	subjLeader = regexp.MustCompile(fmt.Sprintf(`(?i)^(?:(?:%s)*%s)`,
+		subjBlob, subjReFwd))
+	// subj-trailer    = "(fwd)" / WSP
+	subjTrailer = regexp.MustCompile(`(?i)\(fwd\)$`)
+)
+
+// Steps 2-4 in RFC Section 2.1
+func replaceArtifacts(subject string, isReplyFwd *bool) string {
+	// (2) Remove all trailing text of the subject that matches the
+	// subj-trailer ABNF; repeat until no more matches are possible.
+	for {
+		noTrail := strings.TrimSuffix(subject, " ")
+		if subjTrailer.MatchString(noTrail) {
+
+			noTrail = subjTrailer.ReplaceAllString(noTrail, "")
+			*isReplyFwd = true
+		}
+		if subject == noTrail {
+			break
+		}
+		subject = noTrail
+	}
+	return replacePrefix(subject, isReplyFwd)
+}
+
+// Steps 3 & 4 in RFC Section 2.1
+func replacePrefix(subject string, isReplyFwd *bool) string {
+	// Repeat (3) and (4) until no matches remain.
+	for {
+		// (3) Remove all prefix text of the subject that matches the subj-
+		// leader ABNF.
+		noLeader := strings.TrimPrefix(subject, " ")
+		if subjLeader.MatchString(noLeader) {
+			noLeader = subjLeader.ReplaceAllString(noLeader, "")
+			*isReplyFwd = true
+		}
+
+		// (4) If there is prefix text of the subject that matches the subj-
+		// blob ABNF, and removing that prefix leaves a non-empty subj-
+		// base, then remove the prefix text.
+		noBlob := subjBlobPrefix.ReplaceAllString(noLeader, "")
+		if noBlob == "" {
+			subject = noLeader
+			break
+		}
+		if noBlob == subject {
+			break
+		}
+		subject = noBlob
+	}
+	return subject
+}
+
+func GetBaseSubject(subject string, isReplyFwd *bool) (string, error) {
+	var (
+		baseSubject string
+		err         error
+	)
+
+	*isReplyFwd = false
+
+	// (1) Convert any RFC 2047 encoded-words in the subject to [UTF-8]
+	// as described in "Internationalization Considerations".
+	// Convert all tabs and continuations to space.  Convert all
+	// multiple spaces to a single space.
+	dec := new(mime.WordDecoder)
```

Comment:
This will only decode UTF-8. I think it makes sense to expect an already-decoded subject here, because libraries such as go-imap and go-message do it for you.